NBitcoin/Protocol/Behaviors/AddressManagerBehavior.cs

```diff
@@ -136,19 +136,37 @@ protected override void AttachCore()
 
 		void AttachedNode_MessageReceived(Node node, IncomingMessage message)
 		{
+			var payload = message.Message.Payload; 
+			if (payload is SendAddrV2Payload)
+			{
+				node.PreferAddressV2 = true;
+				return;
+			}
+			
 			if ((Mode & AddressManagerBehaviorMode.Advertize) != 0)
 			{
-				var getaddr = message.Message.Payload as GetAddrPayload;
-				if (getaddr != null)
+				if (payload is GetAddrPayload)
 				{
-					node.SendMessageAsync(new AddrPayload(AddressManager.GetAddr().Take(1000).ToArray()));
+					var addresses = AddressManager.GetAddr().Where(a => a.Endpoint.IsValid());
+					if (node.PreferAddressV2)
+					{
+						var addressesToSend = addresses

```

Comment:
Remove spagetti.

Reply:
Done. Could you take a look now?

Reply:
![image](https://user-images.githubusercontent.com/9156103/102080640-19032480-3e0f-11eb-9b23-9b8efc2b131a.png)

Reply:
GJ